File: app/lib/runs/progress.ts
import { z } from "zod"

import { runErrorCode, type ReportRun, type RunStatus } from "@/lib/db/schema"

/**
 * The progress callback's boundary schema and its **pure** decision function
 * (Requirement 38).
 *
 * **Pure, and deliberately not `server-only`.** No I/O, no clock, no environment,
 * no secret: a parsed body and the row's current state go in, and either a
 * rejection or the exact set of columns to write comes out. The route does the
 * three things that are genuinely about the request — validate the token, read the
 * row, apply the write — and nothing else.
 *
 * The split exists because Requirement 38 is mostly a *decision table*, and the
 * table has cases that are easy to get subtly wrong and expensive to reach through
 * a route handler: a repeat of a terminal status must be refused (38.8), a repeat of
 * a **non**-terminal status must be accepted and must still write the counts
 * (38.13), a lower `current` for the same phase must leave the counts alone while
 * applying the rest (38.14), and `TIMEOUT` must be refused outright (38.11). Each
 * of those is one line here and one assertion in the test file.
 *
 * ## `current` on the wire, `progress_current` in the column, `done` in the event
 *
 * The callback field is named **`current`** while the emitted SSE `progress` event's
 * field is named **`done`**. That is deliberate and renames nothing: the callback
 * names a column, the event names a field of the declared vocabulary
 * (Requirement 14.8), and the relay maps `progress_current → done` when it emits
 * (Requirement 40.15). Three names for one number is worse than it sounds only if
 * the mapping is implicit; it is stated here, in `progressColumns`, and in the
 * relay.
 */

// --- The schema -------------------------------------------------------------

/**
 * The phases the **agent** may present.
 *
 * `queued` and `claimed` are absent because the reaper owns them: an agent
 * presenting `claimed` is claiming to have done the claiming. `compiling`,
 * `rendering` and `verifying` are absent because this spec does not drive them
 * (Requirement 36.2) — a callback naming one would be refused by the transition
 * table anyway, and refusing it at the schema says so with a field path.
 *
 * Mirrors `AGENT_PHASES` in `agent/src/reporting_agent/progress.py`.
 */
export const AGENT_PHASES = ["collecting", "completed", "failed"] as const

export type AgentPhase = (typeof AGENT_PHASES)[number]

/** An `error_message` long enough to be useful and bounded enough to store. */
export const MAX_ERROR_MESSAGE_LENGTH = 2000

/** A progress label is a badge — "Inventory", "Metrics" — not a sentence. */
export const MAX_PROGRESS_LABEL_LENGTH = 64

/**
 * `POST /api/internal/runs/[runId]/progress` (Requirements 38.1, 38.7, 38.12).
 *
 * There is **no token field**, and its absence is Requirement 38.2 expressed as a
 * schema: the token travels in the `X-Rpt-Progress-Token` header, never in the
 * request target and never in the body. `.strict()` means a body that carries one
 * anyway is *rejected* rather than having it quietly ignored — which matters,
 * because a caller that put a credential in a body it believed was accepted would
 * have no signal that it had done so.
 *
 * `error_code` is drawn from the Postgres enum's own values rather than a restated
 * list, so a code the column cannot hold can never reach the write. `TIMEOUT` is
 * *accepted by the schema* and refused by {@link decideTransition} — deliberately,
 * so the refusal carries the reason "the reaper is the only writer of TIMEOUT"
 * rather than an anonymous enum mismatch on a field path.
 */
export const progressCallbackSchema = z
  .object({
    run_id: z.string().min(1),
    phase: z.enum(AGENT_PHASES),
    error_code: z.enum(runErrorCode.enumValues).optional(),
    error_message: z.string().max(MAX_ERROR_MESSAGE_LENGTH).optional(),
    /** The snapshot's `content_hash`: 64 lowercase hex. */
    snapshot_id: z
      .string()
      .regex(/^[0-9a-f]{64}$/)
      .optional(),
    resource_count: z.number().int().nonnegative().optional(),
    gap_count: z.number().int().nonnegative().optional(),
    /** Requirement 38.1 — the in-flight count. Named `current`; see the docstring. */
    current: z.number().int().nonnegative().optional(),
    total: z.number().int().positive().optional(),
    label: z.string().max(MAX_PROGRESS_LABEL_LENGTH).optional(),
  })
  .strict()

export type ProgressCallback = z.output<typeof progressCallbackSchema>

// --- The decision -----------------------------------------------------------

/**
 * Why a callback was refused.
 *
 * Every one of these resolves to the **same** HTTP response — a 404 with a fixed
 * body — because Requirement 38.6 requires one response identical for a bad token
 * and an unknown run, and widening that to every rejection is what stops the
 * endpoint from being an oracle for run state. A caller cannot learn from the
 * response whether the run exists, what status it carries, or which of these
 * happened.
 *
 * The reason is still carried, as a value, so the **server log** can say which. That
 * is the only consumer.
 */
export type TransitionRejection =
  | "terminal_row"
  | "unreachable_target"
  | "timeout_reserved"
  | "error_code_not_permitted"
  | "missing_error_code"

/**
 * The columns to write, or a rejection.
 *
 * `write` is a partial row rather than a full one, and the fields it *omits* are as
 * load-bearing as the ones it sets: a same-status refresh omits `status` entirely
 * (Requirement 38.13), and an out-of-order `current` omits all three progress
 * columns (Requirement 38.14). Omission means "leave the column alone", which is a
 * thing a partial `UPDATE ... SET` expresses exactly and a full row cannot.
 */
export type TransitionDecision =
  | { readonly ok: false; readonly rejection: TransitionRejection }
  | {
      readonly ok: true
      /** `true` when `status` changes; `false` for a same-phase refresh. */
      readonly changesStatus: boolean
      readonly write: ProgressWrite
    }

/**
 * Exactly the columns a progress callback may write.
 *
 * `updatedAt` is absent because the caller sets it unconditionally — Requirement
 * 36.3 says every write that changes another column refreshes it, and leaving it
 * out of this type means there is no path on which a decision could forget.
 * `progressTokenHash` is absent because Requirement 38.12 forbids writing any
 * column derived from the presented token: the type is where that is enforced,
 * rather than a rule somebody has to remember while writing the statement.
 */
export type ProgressWrite = {
  readonly status?: RunStatus
  readonly phaseDeadline?: Date | null
  readonly errorCode?: (typeof runErrorCode.enumValues)[number] | null
  readonly errorMessage?: string | null
  readonly snapshotId?: string
  readonly resourceCount?: number
  readonly gapCount?: number
  readonly progressCurrent?: number | null
  readonly progressTotal?: number | null
  readonly progressLabel?: string | null
}

/** The fields of the row this decision reads. A `Pick`, so its inputs are its type. */
export type ProgressRowState = Pick<
  ReportRun,
  "status" | "progressCurrent" | "progressTotal" | "progressLabel"
>

/** The `error_message` recorded when a `failed` callback carried none. */
export const DEFAULT_FAILURE_MESSAGE =
  "The run failed. The runtime reported no message with the failure."

/**
 * Decide what one callback writes (Requirements 38.7, 38.8, 38.10–38.14).
 *
 * The order of the checks below is the order the requirements force, and each early
 * return is a case that must write **nothing**:
 *
 *  1. **A terminal row rejects everything** (38.8), including a repeat of the
 *     terminal status it already carries. Checked first, so a `failed` row cannot be
 *     reopened by a late `collecting` callback and a `completed` row cannot have its
 *     counts rewritten. This is the guard that makes a delivered run's audit row
 *     final.
 *  2. **`TIMEOUT` is refused outright** (38.11). The reaper is its only writer,
 *     because a timed-out run's container may already be gone — so a callback
 *     presenting it is either confused or forged, and in both cases it is claiming
 *     to have observed something it cannot have.
 *  3. **The target must be `{current} ∪ DRIVEN[current]`** (38.10), which is what
 *     stops a replayed or out-of-order callback from moving a run backwards.
 *  4. **A `failed` target's code must be one the agent may present** (38.11).
 *  5. Then the write is assembled, differently for the terminal and non-terminal
 *     cases (38.7, 38.12, 38.13, 38.14).
 *
 * `dependencies` are the two things this function cannot know on its own — the
 * transition table and the phase budget — passed in rather than imported so this
 * module stays free of `server-only` and the table has exactly one definition,
 * in `lib/runs/state.ts`.
 */
export function decideTransition(
  row: ProgressRowState,
  callback: ProgressCallback,
  now: Date,
  dependencies: {
    /** `{current} ∪ DRIVEN[current]`, from `lib/runs/state.ts`. */
    readonly acceptedTargets: (status: RunStatus) => readonly RunStatus[]
    /** The entered phase's budget, from `lib/runs/state.ts`. */
    readonly phaseDeadlineFor: (status: RunStatus, now: Date) => Date | null
    /** The codes the agent may present on a `failed` transition. */
    readonly agentErrorCodes: ReadonlySet<string>
    /** The codes only the app writes — `TIMEOUT`, `SECRET_UNREADABLE`. */
    readonly appWrittenCodes: ReadonlySet<string>
  }
): TransitionDecision {
  const target: RunStatus = callback.phase

  // 1 — Requirement 38.8. `acceptedTargets` returns `[]` for a terminal row, so
  //     step 3 would also refuse this — but the rejection reason differs and the
  //     log line is the only place it shows, so the case is named.
  if (dependencies.acceptedTargets(row.status).length === 0) {
    return { ok: false, rejection: "terminal_row" }
  }

  // 2 — Requirement 38.11, checked before the transition table so the reason is
  //     "TIMEOUT is reserved" rather than "that target is unreachable".
  if (
    callback.error_code !== undefined &&
    dependencies.appWrittenCodes.has(callback.error_code)
  ) {
    return { ok: false, rejection: "timeout_reserved" }
  }

  // 3 — Requirement 38.10.
  if (!dependencies.acceptedTargets(row.status).includes(target)) {
    return { ok: false, rejection: "unreachable_target" }
  }

  // 4 — Requirement 38.11's other half, and the CHECK constraint's precondition:
  //     a `failed` row must carry a code, and the code must be one of the set.
  if (target === "failed") {
    if (callback.error_code === undefined) {
      return { ok: false, rejection: "missing_error_code" }
    }
    if (!dependencies.agentErrorCodes.has(callback.error_code)) {
      return { ok: false, rejection: "error_code_not_permitted" }
    }
  }

  const changesStatus = target !== row.status

  // Requirement 38.12 — a terminal transition records its terminal fields and
  // clears `phase_deadline` **together with** all three progress columns, so a
  // terminal row carries no stale in-flight count.
  if (target === "failed") {
    return {
      ok: true,
      changesStatus,
      write: {
        status: "failed",
        errorCode: callback.error_code,
        errorMessage: callback.error_message ?? DEFAULT_FAILURE_MESSAGE,
        phaseDeadline: null,
        progressCurrent: null,
        progressTotal: null,
        progressLabel: null,
      },
    }
  }

  if (target === "completed") {
    return {
      ok: true,
      changesStatus,
      write: {
        status: "completed",
        // Explicitly `null`, not merely unset. The CHECK forbids a `completed` row
        // from carrying a code, and writing the pair together means the completed
        // row is written whole rather than relying on the previous status never
        // having recorded one.
        errorCode: null,
        errorMessage: null,
        // Requirement 38.12 names these three as the terminal fields for
        // `completed`. Each is written only when presented: a callback that
        // omitted `snapshot_id` should not blank a value already recorded.
        ...(callback.snapshot_id === undefined
          ? {}
          : { snapshotId: callback.snapshot_id }),
        ...(callback.resource_count === undefined
          ? {}
          : { resourceCount: callback.resource_count }),
        ...(callback.gap_count === undefined
          ? {}
          : { gapCount: callback.gap_count }),
        phaseDeadline: null,
        progressCurrent: null,
        progressTotal: null,
        progressLabel: null,
      },
    }
  }

  // A non-terminal target. Requirement 38.7 for a transition, 38.13 for a
  // same-status refresh — and the two differ **only** in whether `status` is set.
  // The deadline is refreshed either way, which is what stops a phase that is
  // making visible progress from being reaped for taking a while.
  return {
    ok: true,
    changesStatus,
    write: {
      ...(changesStatus ? { status: target } : {}),
      phaseDeadline: dependencies.phaseDeadlineFor(target, now),
      ...progressColumns(row, callback, target),
    },
  }
}

/**
 * The three progress columns a non-terminal callback writes, or `{}`.
 *
 * `{}` — leave all three alone — in exactly one case, Requirement 38.14: the
 * presented `current` is **below** the `progress_current` already stored while the
 * row's `status` already equals the presented target. Requirement 14.8 requires
 * successive `done` values for one step to be non-decreasing, and the reporter's
 * single retry can land out of order, so the **row** enforces monotonicity rather
 * than trusting the caller to.
 *
 * The `status` equality is part of the condition, not incidental: a *transition*
 * into a phase legitimately resets the count to zero, and treating that as an
 * out-of-order arrival would freeze the bar at the previous phase's total.
 *
 * All three move together. Writing a new `current` while keeping an old `total`
 * would produce `847 / 200`, and writing a new `total` while keeping an old
 * `current` would make the bar jump backwards — so the guard covers the triple
 * rather than the one field it is named for.
 */
function progressColumns(
  row: ProgressRowState,
  callback: ProgressCallback,
  target: RunStatus
): ProgressWrite {
  const sameStatus = row.status === target

  if (
    sameStatus &&
    callback.current !== undefined &&
    row.progressCurrent !== null &&
    callback.current < row.progressCurrent
  ) {
    return {}
  }

  // Requirement 38.7/38.13 — "each of `progress_current`, `progress_total` and
  // `progress_label` the request presents". A field the callback omitted is
  // written as `null` rather than left alone, so a phase that stops carrying a
  // countable unit of work clears the bar instead of leaving the previous phase's
  // numbers on screen. `null` for either count is what makes the relay emit no
  // `progress` event at all (Requirement 40.14).
  return {
    progressCurrent: callback.current ?? null,
    progressTotal: callback.total ?? null,
    progressLabel: callback.label ?? null,
  }
}

File: agent/src/reporting_agent/progress.py
"""The progress callback — how the agent advances its own run state (Req 38.1–38.4, 38.15).

At every phase transition the runtime fires a short POST to the `progress_url` the
invoke context supplied. Four or five tiny independent requests per run, none
long-lived, none able to time out the way a twelve-minute stream can. Postgres is the
state machine; this module is the only thing in the agent that writes to it.

Four properties are load-bearing, and each one is a decision rather than a detail.

**The token travels in a header** — `X-Rpt-Progress-Token` — never in the request
target and never in the body (Req 38.2, 15.7). It is a credential, not a correlation
id: it authorizes writes to the run state machine, so a leaked token lets someone mark
a run `completed`. A URL is copied into every intermediary access log on the path, so a
token in a query string is a token in somebody else's log file. :func:`_build_body`
therefore constructs the body from a fixed field list and drops any token-shaped key a
caller passes, and nothing here ever appends to `progress_url`.

**Nothing raised here can end a run** (Req 38.4). A run that dies because it could not
report its own progress is the worst of both designs: the work succeeded and the row
says otherwise. Every failure — a timeout, a non-success status, an exception from the
transport — is retried at most once, logged with the token excluded, and then abandoned
(Req 38.3). The Reaper's deadline sweep is the backstop for a callback that never
landed.

**Intermediate transitions are fire-and-forget; the terminal transition is awaited.**
The asymmetry is deliberate. Losing `collecting` costs a stale progress display and the
next transition corrects it. Losing the terminal callback costs a **false `TIMEOUT` on a
run that actually succeeded**, and the container is about to exit, so there is no later
transition to correct it. Awaiting the terminal callback bounds shutdown by
`PROGRESS_TIMEOUT_S * PROGRESS_MAX_ATTEMPTS` — ten seconds worst case — and keeps the
Reaper as the backstop rather than the primary path.

**In-phase progress is throttled to one callback per 5 seconds per phase**
(`PROGRESS_THROTTLE_S`, Req 38.15), with both guards stated positively:

* **every phase transition is sent at the instant it occurs**, irrespective of the
  limit, and
* **the terminal callback is always sent**, irrespective of the limit.

A 200-resource month folds many batches. Posting per folded batch would turn the
design's "four or five tiny requests per run" budget into hundreds of real HTTP
requests against the app, for a counter the UI reads at a two-second poll anyway.
Throttling the *count* while exempting the *transitions* keeps the callback path
bounded without ever delaying the write that moves the state machine.

The clock and the HTTP transport are injected, so the throttle and the retry are
testable over simulated time against a fake transport, with no network and no sleeping.
"""

from __future__ import annotations

import asyncio
import logging
import time
from collections.abc import Callable, Mapping
from typing import Any, Final, Protocol, runtime_checkable

from reporting_agent.errors import APP_WRITTEN_CODES, ROW_ERROR_CODES
from reporting_agent.redaction import presence_marker, register_secrets, scrub, scrub_exception

__all__ = [
    "AGENT_PHASES",
    "DOCUMENT_PHASES",
    "PROGRESS_MAX_ATTEMPTS",
    "PROGRESS_THROTTLE_S",
    "PROGRESS_TIMEOUT_S",
    "TERMINAL_PHASES",
    "TOKEN_HEADER",
    "HttpxProgressTransport",
    "ProgressReporter",
    "ProgressTransport",
]

logger = logging.getLogger(__name__)

PROGRESS_TIMEOUT_S: Final[float] = 5.0
"""Req 38.3 — a callback that has not completed within this is a failure."""

PROGRESS_MAX_ATTEMPTS: Final[int] = 2
"""One retry, and no more (Req 38.3). Two attempts at 5 seconds is what bounds the
awaited terminal callback to ten seconds, so there is deliberately **no backoff sleep**
between them: a delay here is time added to container shutdown, and the failure this
retry covers is a dropped connection rather than a busy server."""

PROGRESS_THROTTLE_S: Final[float] = 5.0
"""Req 38.15 — at most one *in-phase progress* callback per phase per 5 seconds."""

TOKEN_HEADER: Final[str] = "X-Rpt-Progress-Token"
"""Req 38.2 — where the run-scoped token is presented, and the only place it appears."""

TERMINAL_PHASES: Final[frozenset[str]] = frozenset({"completed", "failed"})

DOCUMENT_PHASES: Final[frozenset[str]] = frozenset(
    {"compiling", "rendering", "verifying"}
)
"""The three phases the document pipeline drives (Req 41.3).

Added when their transitions landed in `lib/runs/state.ts`'s `DRIVEN` table. Presenting a
phase the endpoint's table does not admit spends a request to be refused, so this set and
that table are two spellings of one fact — and the integration tests in task 11.6 assert
over all `(current, target)` pairs, which is what keeps them one fact rather than two.
"""

AGENT_PHASES: Final[frozenset[str]] = (
    frozenset({"collecting"}) | DOCUMENT_PHASES | TERMINAL_PHASES
)
"""The phases the **agent** may present, matching `progressCallbackSchema`'s enum.

`queued` and `claimed` are absent because the Reaper owns them — an agent presenting
`claimed` is claiming to have done the claiming. `TIMEOUT` is likewise the Reaper's: the
agent may already be gone when a deadline elapses, so a transition carrying it is a request
guaranteed to be refused (Req 41.5).
"""

# The codes the agent may present on a `failed` transition. `TIMEOUT` and
# `SECRET_UNREADABLE` are written by the app — the Reaper's sweep and the tick's
# decryption failure — and the endpoint rejects a presented `TIMEOUT` outright
# (Req 38.11), so a callback carrying one is a request guaranteed to be refused.
_AGENT_ERROR_CODES: Final[frozenset[str]] = ROW_ERROR_CODES - APP_WRITTEN_CODES

# Anything whose name looks like a credential is dropped from the body rather than
# trusted not to have been passed (Req 38.2, 15.7).
_FORBIDDEN_BODY_KEYS: Final[frozenset[str]] = frozenset(
    {"progress_token", "progresstoken", "token", "client_secret", "clientsecret"}
)

# The endpoint parses the body with zod, and a parse failure refuses the **whole**
# callback — including the phase transition it carries. So a field that cannot be valid
# is repaired or dropped here rather than sent: losing one label is a cosmetic
# regression, while losing the transition it travelled with costs a false `TIMEOUT`.
_MAX_ERROR_MESSAGE_CHARS: Final[int] = 2000
_MAX_LABEL_CHARS: Final[int] = 64

_NON_TERMINAL_BODY_FIELDS: Final[tuple[str, ...]] = ("current", "total", "label")
_FAILED_BODY_FIELDS: Final[tuple[str, ...]] = ("error_code", "error_message")
_COMPLETED_BODY_FIELDS: Final[tuple[str, ...]] = (
    "snapshot_id",
    "resource_count",
    "gap_count",
)


def _coerce(name: str, value: Any) -> Any | None:
    """Repair a body field the endpoint would refuse, or drop it.

    Over-long text is truncated because the message still carries its useful prefix. A
    count outside its declared range is dropped rather than clamped: a negative count is
    a bug in the caller, and inventing a plausible substitute for it would put a wrong
    number on the row this product exists to keep honest.
    """
    if name == "error_message":
        return str(value)[:_MAX_ERROR_MESSAGE_CHARS]
    if name == "label":
        return str(value)[:_MAX_LABEL_CHARS]

    if name in ("current", "total", "resource_count", "gap_count"):
        # `bool` is a subclass of `int`, and `True` is not a resource count.
        if isinstance(value, bool) or not isinstance(value, int):
            logger.warning("%s=%r is not an integer; dropped from the callback body.", name, value)
            return None
        floor = 1 if name == "total" else 0
        if value < floor:
            logger.warning(
                "%s=%d is below the minimum the progress endpoint accepts (%d); dropped "
                "from the callback body so the transition it travels with still lands.",
                name,
                value,
                floor,
            )
            return None

    return value


@runtime_checkable
class ProgressTransport(Protocol):
    """The one HTTP operation this module performs, injected so it can be faked.

    Returns the response status code. Anything outside 2xx is a failure, and so is
    raising — the caller treats the two identically, because from the run's point of
    view they are the same event: the transition did not land.
    """

    async def post_json(
        self,
        url: str,
        *,
        body: Mapping[str, Any],
        headers: Mapping[str, str],
        timeout: float,
    ) -> int: ...


class HttpxProgressTransport:
    """`ProgressTransport` over `httpx.AsyncClient`.

    The client is built on first use rather than in `__init__`, so constructing a
    reporter opens no connection pool — a prompt invocation that never reports a phase
    pays nothing. `client` is injectable for the same reason it is elsewhere in this
    package: so a test can exercise the real code path without a socket.
    """

    def __init__(self, client: Any | None = None) -> None:
        self._client = client

    async def post_json(
        self,
        url: str,
        *,
        body: Mapping[str, Any],
        headers: Mapping[str, str],
        timeout: float,
    ) -> int:
        if self._client is None:
            import httpx

            self._client = httpx.AsyncClient()

        response = await self._client.post(
            url, json=dict(body), headers=dict(headers), timeout=timeout
        )
        return int(response.status_code)

    async def aclose(self) -> None:
        client = self._client
        self._client = None
        if client is not None and hasattr(client, "aclose"):
            await client.aclose()


class ProgressReporter:
    """Sends this run's phase transitions to the app's progress endpoint.

    One instance per invocation, built from the invoke `context`. When `progress_url` or
    `progress_token` is absent — a chat prompt carries no run — the reporter is
    **disabled** and every method is a no-op, so no caller needs to branch on whether
    this invocation has a run row behind it.
    """

    def __init__(
        self,
        *,
        progress_url: str | None,
        progress_token: str | None,
        run_id: str | None,
        transport: ProgressTransport | None = None,
        clock: Callable[[], float] = time.monotonic,
    ) -> None:
        self._url = (progress_url or "").strip()
        self._token = progress_token or ""
        self._run_id = (run_id or "").strip()
        self._transport: ProgressTransport = (
            transport if transport is not None else HttpxProgressTransport()
        )
        self._clock = clock

        self.enabled: bool = bool(self._url and self._token and self._run_id)

        # The phase the row is believed to carry. A `report` naming a different phase is
        # a *transition* and is exempt from the throttle; one naming this phase is an
        # in-phase refresh and is not.
        self._current_phase: str | None = None
        self._last_sent_at: dict[str, float] = {}
        self._pending: set[asyncio.Task[None]] = set()
        self._terminal_sent = False

        # Belt and braces. `main.py` registers the token when it parses the context
        # (Req 15.1) and registration is idempotent, so this costs nothing there — but
        # it means a log line from *this* module cannot carry the token even if a caller
        # built a reporter without going through that path.
        if self._token:
            register_secrets((self._token,))

        # Req 38.2 / 15.7: the token must not be in the request target. This module never
        # appends to `progress_url`, so the only way it could be is if the caller was
        # handed a URL that already contains it — which is a caller bug worth naming,
        # with the value itself excluded from the message.
        if self._token and self._token in self._url:
            logger.warning(
                "progress_url contains the run-scoped progress token; the token must "
                "travel in the %s header only. Continuing with the URL as supplied.",
                TOKEN_HEADER,
            )

    # --- the two public operations ---------------------------------------------------

    async def report(
        self,
        phase: str,
        *,
        current: int | None = None,
        total: int | None = None,
        label: str | None = None,
    ) -> None:
        """Report a **non-terminal** phase, fire-and-forget. Never raises (Req 38.4).

        Returns as soon as the request is scheduled, so a collector fold is never
        blocked on the app's response. `current`, `total` and `label` are carried where
        the entered phase has a countable unit of work (Req 38.1), which is what lets the
        row feed a determinate bar rather than a spinner that runs for four minutes.

        A transition to this phase is sent at the instant it occurs; a refresh inside the
        phase is subject to `PROGRESS_THROTTLE_S` (Req 38.15).
        """
        if not self.enabled or self._terminal_sent:
            return

        try:
            if phase in TERMINAL_PHASES:
                # Never drop a terminal callback on the floor: losing it costs a false
                # TIMEOUT on a successful run. Route it to the awaited path instead.
                logger.warning(
                    "report(%r) names a terminal phase; sending it through the awaited "
                    "terminal path instead of fire-and-forget.",
                    phase,
                )
                await self.report_terminal(phase)
                return

            if phase not in AGENT_PHASES:
                # Nothing outside the endpoint's enum can be accepted, so spending a
                # request to be refused buys nothing. Name it in a log line instead.
                logger.warning(
                    "phase %r is not a phase the agent may present; nothing sent.", phase
                )
                return

            if not self._admit(phase):
                return

            body = self._build_body(phase, current=current, total=total, label=label)
            task = asyncio.create_task(
                self._deliver(phase, body), name=f"progress-callback-{phase}"
            )
            self._pending.add(task)
            task.add_done_callback(self._pending.discard)
        except Exception as exc:  # Req 38.4: this may not end the run
            logger.warning(
                "progress callback for phase %r could not be scheduled; the run "
                "continues and the reaper is the backstop: %s",
                phase,
                scrub_exception(exc),
            )

    async def report_terminal(
        self,
        phase: str,
        *,
        error_code: str | None = None,
        error_message: str | None = None,
        snapshot_id: str | None = None,
        resource_count: int | None = None,
        gap_count: int | None = None,
    ) -> None:
        """Report a terminal phase and **wait for it**. Never raises (Req 38.4).

        Always sent, irrespective of the throttle (Req 38.15). Awaited because the
        container is about to exit: an unsent terminal callback is a run that succeeded
        and will be failed as `TIMEOUT` by the Reaper.

        Terminal fields are routed by phase, as Req 38.12 declares them — `error_code`
        and `error_message` for `failed`, `snapshot_id`, `resource_count` and `gap_count`
        for `completed`. A field belonging to the other phase is dropped, and no
        `current`/`total`/`label` is presented at all, because a terminal transition
        clears those columns.
        """
        if not self.enabled:
            return

        try:
            if phase not in TERMINAL_PHASES:
                logger.warning(
                    "report_terminal(%r) names a non-terminal phase; nothing sent.", phase
                )
                return

            body = self._build_body(
                phase,
                error_code=error_code,
                error_message=error_message,
                snapshot_id=snapshot_id,
                resource_count=resource_count,
                gap_count=gap_count,
            )

            self._current_phase = phase
            self._last_sent_at[phase] = self._clock()
            self._terminal_sent = True

            await self._deliver(phase, body)
        except Exception as exc:  # Req 38.4
            logger.warning(
                "terminal progress callback for phase %r failed; the run's outcome "
                "stands and the reaper is the backstop: %s",
                phase,
                scrub_exception(exc),
            )
        finally:
            await self.aclose()

    async def aclose(self) -> None:
        """Abandon any still-in-flight intermediate callback. Never raises.

        Called after the terminal callback. An intermediate transition still in flight at
        that point is *superseded* — the endpoint rejects every transition on a terminal
        row — so cancelling is both correct and instant, which is what keeps container
        shutdown bounded by the terminal callback alone. It also stops asyncio warning
        about a pending task destroyed at loop close.
        """
        pending = tuple(self._pending)
        self._pending.clear()
        for task in pending:
            task.cancel()
        if pending:
            await asyncio.gather(*pending, return_exceptions=True)

    # --- the throttle ----------------------------------------------------------------

    def _admit(self, phase: str) -> bool:
        """May this non-terminal callback be sent now (Req 38.15)?

        Decided **synchronously**, before the request is scheduled, so the limit bounds
        HTTP requests rather than tasks. `_last_sent_at` records the decision instant,
        not the completion instant: a slow response must not widen the window and let a
        second request through early.
        """
        now = self._clock()

        if phase != self._current_phase:
            # A transition. Sent at the instant it occurs, irrespective of the limit.
            self._current_phase = phase
            self._last_sent_at[phase] = now
            return True

        last = self._last_sent_at.get(phase)
        if last is not None and (now - last) < PROGRESS_THROTTLE_S:
            return False

        self._last_sent_at[phase] = now
        return True

    # --- the body --------------------------------------------------------------------

    def _build_body(self, phase: str, **fields: Any) -> dict[str, Any]:
        """Build the callback body from a fixed field list.

        Constructed by allow-list rather than by copying a caller's mapping, which is
        what makes "the token is never in the body" (Req 38.2) structural: there is no
        field name here that could carry it, and a token-shaped key is dropped with a
        warning rather than forwarded.
        """
        body: dict[str, Any] = {"run_id": self._run_id, "phase": phase}

        if phase == "failed":
            permitted = _FAILED_BODY_FIELDS
        elif phase == "completed":
            permitted = _COMPLETED_BODY_FIELDS
        else:
            permitted = _NON_TERMINAL_BODY_FIELDS

        for name, value in fields.items():
            if name.lower() in _FORBIDDEN_BODY_KEYS:
                logger.warning(
                    "field %r was offered for a progress callback body and has been "
                    "dropped; a credential travels in the %s header only.",
                    name,
                    TOKEN_HEADER,
                )
                continue
            if value is None or name not in permitted:
                continue
            if name == "error_code" and str(value) not in _AGENT_ERROR_CODES:
                # Req 38.11 — the reaper is the only writer of TIMEOUT, and a `failed`
                # transition carrying an out-of-set code is refused. Dropping the field
                # here names the mistake in a log line instead of spending a request.
                logger.warning(
                    "error_code %r is not a code the agent may present; it has been "
                    "dropped from the progress callback body.",
                    value,
                )
                continue

            coerced = _coerce(name, value)
            if coerced is None:
                continue
            body[name] = coerced

        return body

    # --- delivery --------------------------------------------------------------------

    async def _deliver(self, phase: str, body: Mapping[str, Any]) -> None:
        """POST `body`, retrying at most once. Never raises (Req 38.3, 38.4)."""
        headers = {TOKEN_HEADER: self._token}

        for attempt in range(1, PROGRESS_MAX_ATTEMPTS + 1):
            try:
                status = await asyncio.wait_for(
                    self._transport.post_json(
                        self._url,
                        body=body,
                        headers=headers,
                        timeout=PROGRESS_TIMEOUT_S,
                    ),
                    timeout=PROGRESS_TIMEOUT_S,
                )
            except asyncio.CancelledError:
                # Cooperative cancellation from `aclose`, not a failure. Propagate.
                raise
            except TimeoutError:
                # Built-in `TimeoutError` subclasses `OSError`, so an `except OSError`
                # added above or beside this arm would swallow every timeout first and
                # report it as a transport failure — losing the "slow" signal that is the
                # only thing distinguishing this arm from the fallback below. On 3.11+
                # `asyncio.TimeoutError is TimeoutError`, so this one clause covers both.
                self._log_failure(phase, attempt, detail="did not complete within 5s")
            except Exception as exc:  # Req 38.4
                self._log_failure(phase, attempt, detail=scrub_exception(exc))
            else:
                if 200 <= status < 300:
                    return
                self._log_failure(phase, attempt, detail=f"returned status {status}")

        logger.warning(
            "progress callback for phase %r abandoned after %d attempts; the run "
            "continues and the reaper is the backstop for a callback that never "
            "arrived.",
            phase,
            PROGRESS_MAX_ATTEMPTS,
        )

    def _log_failure(self, phase: str, attempt: int, *, detail: str) -> None:
        """Record one failed attempt with the token excluded (Req 38.3, 15.7).

        The token is rendered as a presence marker, and the URL is passed through the
        redaction guard, so neither the message nor the arguments can carry the value —
        even if a caller supplied a URL containing it.
        """
        logger.warning(
            "progress callback attempt %d/%d for phase %r to %s failed (token %s): %s",
            attempt,
            PROGRESS_MAX_ATTEMPTS,
            phase,
            scrub(self._url),
            presence_marker(self._token),
            detail,
        )
